### PR TITLE
Add Hotswap

### PR DESCRIPTION
**Pull request by cuongph87**
Added Hotswap.

### CODE REVIEW

File: lib/src/sep/0099/hotswap.dart [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+  final String replacementAsset;
+  final double minimumAmount;
+  final Uri endpointUrl;
+  Dio? dio;
+
+  /// Constructor
+  /// - Parameter url: Endpoint to be used for fetching the information from as well as sending the transaction to.
+  /// - Parameter dio: (Optional) the Dio client.
+
+  static Future<Hotswap> fromEndpointUrl(
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:30]: I think we should try to follow the appraoch in the rest of the SDK with a `fromDomain`.

File: lib/src/sep/0099/hotswap.dart [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+      destinationAddress: json['receivables_address'] as String,
+      currentAsset: json['you_send_asset'] as String,
+      replacementAsset: json['we_send_asset'] as String,
+      minimumAmount: double.parse(json['minimum_amount'] as String),
+      endpointUrl: url,
+      dio: dio,
+    );
+  }
+
+  Future<Transaction> getHotswapSignedTransaction(
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:52]:
```suggestion
  Future<Transaction> getSignedHotswapTransaction(
```

File: lib/src/sep/0099/hotswap.dart [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  final Uri endpointUrl;
+  Dio? dio;
+
+  /// Constructor
+  /// - Parameter url: Endpoint to be used for fetching the information from as well as sending the transaction to.
+  /// - Parameter dio: (Optional) the Dio client.
+
+  static Future<Hotswap> fromEndpointUrl(
+    Uri url, {
+    Dio? dio,
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:32]: I think in the rest of the SDK they use `http.Client` right?

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

File: lib/src/sep/0099/hotswap.dart [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:http/http.dart' as http;
+import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';
+
+class Hotswap {
+  late String _hotswapServerAddress;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart]: We're already in the hotswap class so the context is already clear.
```suggestion
  late String _serverAddress;
```

File: lib/src/sep/0099/hotswap.dart [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  ) async {
+    var account = await _sdk.accounts.account(accountId);
+
+    final hotswapInfo = await info();
+    final txBuilder = TransactionBuilder(account);
+
+    final toAsset = hotswapInfo.toAsset.toAsset();
+    final trustDestinationAssetOperation = ChangeTrustOperationBuilder(
+      toAsset,
+      ChangeTrustOperationBuilder.MAX_LIMIT,
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:69]: Make this an configurable via optional parameter

File: lib/src/sep/0099/hotswap.dart [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+      hotswapHandlerAccountId,
+      fromAsset,
+      fromAssetBalanceObject.balance,
+    ).build();
+
+    // Receive to asset from hotswap server
+    final receiveDestinationAssetOperation = PaymentOperationBuilder(
+      accountId,
+      toAsset,
+      fromAssetBalanceObject.balance,
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:90]:
```suggestion
      // Ensuring a 1:1 exchange
      fromAssetBalanceObject.balance,
```

File: lib/src/sep/0099/hotswap.dart [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+    final transaction = txBuilder.build();
+
+    transaction.sign(
+      signer,
+      _network,
+    );
+
+    var transactionXdr = transaction.toEnvelopeXdrBase64();
+    Uri hotswapUrl =
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:112]:
```suggestion
    var hotswapUrl =
```

File: lib/src/sep/0099/hotswap.dart [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        Util.appendEndpointToUrl(_hotswapServerAddress, 'hotswap').replace(
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:113]:
```suggestion
        Util.appendEndpointToUrl(_hotswapServerAddress, 'hotswap',).replace(
```

File: lib/src/sep/0099/hotswap.dart [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    return HotswapInfoResponse(
+      toAddress: json['receivables_address'] as String,
+      fromAsset: json['you_send_asset'] as String,
+      toAsset: json['we_send_asset'] as String,
+      minimumAmount: double.parse(json['minimum_amount'] as String),
+    );
+  }
+}
+
+extension StringX on String {
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:164]: Can this not be added at a centralised place for extensions?

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

File: lib/src/sep/0099/hotswap.dart (hunk 7 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:112]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 8 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:113]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 6 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:90]: Is it just a comment?

File: lib/src/sep/0099/hotswap.dart (hunk 4 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 5 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:69]: Done.

Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart]: Done.

File: lib/src/sep/0099/hotswap.dart [hunk 10]
```diff
@@ -6,23 +6,24 @@ import 'dart:async';
 import 'dart:convert';
 
 import 'package:http/http.dart' as http;
+import 'package:stellar_flutter_sdk/src/extensions/extensions.dart';
 import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';
 
 class Hotswap {
-  late String _hotswapServerAddress;
+  late final String _serverAddress;
   late http.Client httpClient;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:14]:
```suggestion
  late final http.Client httpClient;
```

File: lib/src/sep/0099/hotswap.dart [hunk 11]
```diff
@@ -6,23 +6,24 @@ import 'dart:async';
...
 
   Hotswap(
-    this._hotswapServerAddress, {
+    String serverAddress, {
     http.Client? httpClient,
   }) {
-    this.httpClient = httpClient == null ? http.Client() : httpClient;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart]: If we create it we also need a dispose for the http client we created manually.

File: lib/src/sep/0099/hotswap.dart [hunk 12]
```diff
@@ -34,42 +35,44 @@ class Hotswap {
...
-    String accountId,
-    KeyPair signer,
-    StellarSDK _sdk,
-    Network _network,
-  ) async {
-    var account = await _sdk.accounts.account(accountId);
+  Future<Transaction> getSignedTransaction({
+    required String accountId,
+    required KeyPair signer,
+    required StellarSDK sdk,
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:58]: Can we not add teh stellar sdk adn the network to the constructors?

File: lib/src/sep/0099/hotswap.dart [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+              (hotswap) => HotswapRoute.fromJson(
+                {
+                  'hotswaps': [hotswap]
+                },
+              ),
+            )
+            .toList(),
+      );
+    } else {
+      // Single hotswap case
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:185]:
```suggestion
      // Single hotswap case (legacy)
      // TODO: Remove as soon as MYKOBO has updated their implementation
```

File: lib/src/sep/0099/hotswap.dart [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+      return _HotswapInfoResponse(
+        hotswapRoutes: [
+          HotswapRoute.fromJson(json),
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:188]: I proposed a change in naming for the properties, so you can't use the same object for parsing.

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

File: lib/src/sep/0099/hotswap.dart (hunk 12 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:58]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 11 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 10 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:14]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 13 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:185]: Done.

File: lib/src/sep/0099/hotswap.dart (hunk 14 above)
Comment (cuongph87) [reply lib/src/sep/0099/hotswap.dart:188]: Done.

File: lib/src/sep/0099/hotswap.dart [hunk 15]
```diff
@@ -11,18 +11,25 @@ import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';
 
 class Hotswap {
   late final String _serverAddress;
-  late http.Client httpClient;
+  late final http.Client httpClient;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:14]:
```suggestion
  late final http.Client _httpClient;
```

File: lib/src/sep/0099/hotswap.dart [hunk 16]
```diff
@@ -11,18 +11,25 @@ import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';
...
+  final StellarSDK sdk;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:15]:
```suggestion
  final StellarSDK _sdk;
```

File: lib/src/sep/0099/hotswap.dart [hunk 17]
```diff
@@ -11,18 +11,25 @@ import 'package:stellar_flutter_sdk/stellar_flutter_sdk.dart';
...
+  final Network network;
```
Comment (jopmiddelkamp) [inline lib/src/sep/0099/hotswap.dart:16]:
```suggestion
  final Network _network;
```

**Review by jopmiddelkamp — CHANGES_REQUESTED** [review]

**Review by jopmiddelkamp — APPROVED** [review]

**Pull request merged**